### PR TITLE
[android] select all ux changes

### PR DESCRIPTION
**Pull request by grzesiekrzaca**
resubmitted for review

### CODE REVIEW

**Comment by aldenml** [comment]
Isn't it possible to solve this at the style level? That way would be more easy to use the idea in other parts (and avoid `PreferenceSlaveLinearLayout`)

**Comment by grzesiekrzaca** [comment]
We can't get the status of the preference in a layout itself (or at least I don't know how to do that) and we cannot assign a style to a specific preference (generic) [even if we could, first problem stands]. I think we would need a custom preference class that is able to change its background based on its checked state. Is that solution preferable?

**Comment by aldenml** [comment]
I might be wrong, but I believe this preferences are some sort of list with an adapter. If that's the case, well crafted selector should do the trick, and that could be reused in "My Files" I guess. Did you researched that route?

**Comment by grzesiekrzaca** [comment]
To hack the adapter for that list hijack getView and check if it not select all? Hmm I can try that tomorrow

File: android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java [hunk 1]
```diff
@@ -17,8 +17,14 @@
 
 package com.frostwire.android.gui.fragments.preference;
 
+import android.support.v4.content.ContextCompat;
```
Comment (gubatron) [inline android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java]: genuine question, is this ok as 'v4', no need for 'v7'? (trying to learn)

File: android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java [hunk 2]
```diff
@@ -141,4 +151,33 @@ private void checkAllEngines(boolean checked) {
             UIUtils.showShortMessage(getView(), R.string.search_preferences_one_engine_checked_always);
         }
     }
+
+    @Override
+    protected RecyclerView.Adapter onCreateAdapter(PreferenceScreen preferenceScreen) {
```
Comment (gubatron) [inline android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java]: oh cool, we're moving to `RecyclerView` already

File: android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java (hunk 1 above)
Comment (aldenml) [inline android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java]: I hope no

**Comment by aldenml** [comment]
@grzesiekrzaca Why do need to setup `android:layout="@layout/view_preference_search_engine"` for each preference?

**Comment by gubatron** [comment]
reviewing functionality (for now) on Android 4.4, 5, 6 and 7.

**Comment by grzesiekrzaca** [comment]
@aldenml 
So that we have one more linear layout (normally there is only one that holds the ripple effect, I'm adding another one that allows me to change the color without destroying ripple effect)

**Comment by gubatron** [comment]
Perhaps this needs a rebase:

This is what I get when I try to install on Android 4.4 (API 19)
```
Error:Execution failed for task ':transformClassesWithDexForBasicDebug'.
> com.android.build.api.transform.TransformException: com.android.ide.common.process.ProcessException: java.util.concurrent.ExecutionException: com.android.dex.DexIndexOverflowException: method ID not in [0, 0xffff]: 65536
```

It's hitting the 65k symbol limit.

Let's see on Android 5.

**Comment by aldenml** [comment]
@gubatron for android 4 you need to disable "Instant Run"

**Comment by gubatron** [comment]
that worked! thanks

**Comment by gubatron** [comment]
works fine on Android 4.4

**Comment by gubatron** [comment]
all good on Android 5.1

File: android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java [hunk 3]
```diff
@@ -141,4 +151,33 @@ private void checkAllEngines(boolean checked) {
...
+            this.checkedDrawableId = checkedDrawableId;
+            this.unCheckedDrawableId = unCheckedDrawableId;
+        }
+
+        @Override
+        public void onBindViewHolder(PreferenceViewHolder holder, int position) {
+            final CheckBoxPreference preference = (CheckBoxPreference) getItem(position);
+
+            preference.onBindViewHolder(holder);
+            if (!preference.getKey().equals("frostwire.prefs.search.preference_category.select_all")) {
```
Comment (gubatron) [inline android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java]: isn't there a `Constants.PREF_KEY_...` value for this?
There should be one, we don't usually have preferences like this hardcoded in the code.
If you find any, let's make sure we use `Constants.PREF_KEY_...` values

Comment (gubatron) [inline android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java]: I've gone ahead and refactored this for you (sending a Pull Request)

However, I'm not sure why it's necessary to have this as a setting, as the state of this checkbox can be set programmatically on the fly if all the search engines available are checked.

**Comment by gubatron** [comment]
All good on Android 6.0.

Sent you a PR with refactors
https://github.com/grzesiekrzaca/frostwire/pull/2

I'll be waiting for an explanation on why the added complexity of a new setting. I think it's unnecessary.

**Comment by gubatron** [comment]
ok, I think the explanation will be that this was the easiest way to add that checkbox in there, instead of adding some hack to the layout.

If that's the case, LGTM @aldenml

**Comment by gubatron** [comment]
@grzesiekrzaca if @aldenml is good with this, let's not just merge.
Let's rebase all this, then squash into a single commit before merging.

**Comment by aldenml** [comment]
Well
1- I don't think `view_preference_search_engine.xml` is necessary. You should be able to craft a good selector and setting it as background in the adapter.
2- Since the key `"frostwire.prefs.search.preference_category.select_all"` is necessary for the lookup, that should be a constant inside this class `SearchFragment` (not in `Constants`). This is a "key reference", not a preference we can access outside the preference UI.
3- You don't need `android.support.v4.content.ContextCompat`
4- I suppose you can make `CheckedAwarePreferenceGroupAdapter` static.

**Comment by gubatron** [comment]
>2- Since the key "frostwire.prefs.search.preference_category.select_all" is necessary for the lookup, that should be a constant inside this class SearchFragment (not in Constants). This is a "key reference", not a preference we can access outside the preference UI.

good point.

**Comment by gubatron** [comment]
Points 2 (constant) and 4 (static class) [addressed here](https://github.com/grzesiekrzaca/frostwire/pull/2/commits/00832dbe1ab6ea76bfd536f9d7edce406b3e4e9e) - (PR for @grzesiekrzaca's branch)

**Comment by grzesiekrzaca** [comment]
@aldenml 
1- well here's the problem:
a) if I change the background with another ripple and a color the hotspot gets deleted - the ripple is started from the top left corner.
```
<ripple xmlns:android="http://schemas.android.com/apk/res/android" android:color="@color/basic_gray_dark">
    <item android:drawable="@color/selected_search_background" />
</ripple>
```
(and another with different color)
I tried to get a touch event and set hotspot there but that doesn't work. 
b) selector won't work because there is no state_checked to be processed in linear layout (this was the reason I had the custom linear layout before)

File: android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java [hunk 4]
```diff
@@ -141,4 +155,32 @@ private void checkAllEngines(boolean checked) {
...
+    }
+
+    private static class CheckedAwarePreferenceGroupAdapter extends PreferenceGroupAdapter {
+        final int checkedDrawableId;
+        final int unCheckedDrawableId;
+        final WeakReference<Activity> activityRef;
+
+        CheckedAwarePreferenceGroupAdapter(Activity activity, PreferenceGroup preferenceGroup, int checkedDrawableId, int unCheckedDrawableId) {
+            super(preferenceGroup);
+            activityRef = Ref.weak(activity);
```
Comment (aldenml) [inline android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java]: Why do we need an activity here?

File: android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java [hunk 5]
```diff
@@ -141,4 +155,32 @@ private void checkAllEngines(boolean checked) {
...
+            this.unCheckedDrawableId = unCheckedDrawableId;
+        }
+
+        @Override
+        public void onBindViewHolder(PreferenceViewHolder holder, int position) {
+            final CheckBoxPreference preference = (CheckBoxPreference) getItem(position);
+            preference.onBindViewHolder(holder);
+            if (!preference.getKey().equals(PREF_KEY_SEARCH_SELECT_ALL) && Ref.alive(activityRef)) {
+                int drawableId = preference.isChecked() ? checkedDrawableId : unCheckedDrawableId;
+                holder.itemView.setBackground(activityRef.get().getResources().getDrawable(drawableId));
```
Comment (aldenml) [inline android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java]: can't the resources be set by id?

**Comment by aldenml** [comment]
@grzesiekrzaca ok, I suppose we can take a deep look at it later. I left a couple of comments that I need you to address before merge

File: android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java [hunk 6]
```diff
@@ -177,9 +172,9 @@ private void checkAllEngines(boolean checked) {
         public void onBindViewHolder(PreferenceViewHolder holder, int position) {
             final CheckBoxPreference preference = (CheckBoxPreference) getItem(position);
             preference.onBindViewHolder(holder);
-            if (!preference.getKey().equals(PREF_KEY_SEARCH_SELECT_ALL) && Ref.alive(activityRef)) {
+            if (!preference.getKey().equals(PREF_KEY_SEARCH_SELECT_ALL)) {
                 int drawableId = preference.isChecked() ? checkedDrawableId : unCheckedDrawableId;
-                holder.itemView.setBackground(activityRef.get().getResources().getDrawable(drawableId));
+                holder.itemView.setBackgroundResource(drawableId);
```
Comment (gubatron) [inline android/src/com/frostwire/android/gui/fragments/preference/SearchFragment.java]: oh cool, forgot about this method

**Pull request merged**